[JSC] Implement `firstDayOfWeek` for `Intl.Locale` info API

sosukesuzuki opened the pull request:
#### 8cd1f6cc395ca82effa2968efe87339d54be2b66
<pre>
[JSC] Implement `firstDayOfWeek` for `Intl.Locale` info API
<a href="https://bugs.webkit.org/show_bug.cgi?id=276943">https://bugs.webkit.org/show_bug.cgi?id=276943</a>

Reviewed by Ross Kirsling.

JSC has implemented the Stage 3 Intl.Locale Info API[1][2]. The proposal includes support for the
&quot;fw&quot; (First day of the week) Unicode extension[3], but it is not currently implemented in JSC.

This patch implements support for the &quot;fw&quot; extension in the Intl.Locale Info API.

[1]: <a href="https://github.com/tc39/proposal-intl-locale-info">https://github.com/tc39/proposal-intl-locale-info</a>
[2]: <a href="https://tc39.es/proposal-intl-locale-info">https://tc39.es/proposal-intl-locale-info</a>
[3]: <a href="https://unicode.org/reports/tr35/#Key_Type_Definitions">https://unicode.org/reports/tr35/#Key_Type_Definitions</a>

* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::weekDayToString):
(JSC::IntlLocale::initializeLocale):
(JSC::IntlLocale::firstDayOfWeek):
* Source/JavaScriptCore/runtime/IntlLocale.h:
* Source/JavaScriptCore/runtime/IntlLocalePrototype.cpp:
(JSC::JSC_DEFINE_CUSTOM_GETTER):

Canonical link: <a href="https://commits.webkit.org/281510@main">https://commits.webkit.org/281510@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/429f0c1c5ea239229d260b661bdbe80a6c7f8ad7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59396 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11916 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63310 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9909 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46393 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10070 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48226 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6976 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61426 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36186 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51409 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29051 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32894 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8662 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8843 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/52488 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54845 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8943 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65043 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58641 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3324 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8857 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55565 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3335 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51405 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55669 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13330 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2761 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80398 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34555 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13934 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35638 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36724 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35383 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/JavaScriptCore/runtime/IntlLocale.cpp:
```diff
@@ -226,6 +226,28 @@ void IntlLocale::initializeLocale(JSGlobalObject* globalObject, JSValue tagValue
 }
 
+// https://tc39.es/proposal-intl-locale-info/#sec-weekday-to-string
+static String weekDayToString(const String fw)
```
- This should be `weekday`, not `weekDay`.
- The param type should be either `const String&` or `StringView`; the latter is preferred, so let's use it.

rkirsling commented on JSTests/stress/intl-locale-info-firstDayOfWeek.js:
```diff
@@ -0,0 +1,51 @@
+        const loc = new Intl.Locale("en", { firstDayOfWeek: weekDay });
+        sameValue(loc.firstDayOfWeek, string);
+    }
+}
```
nit: Let's use `weekday[s]` here too.

rkirsling reviewed (approved):
r=me with comments

rkirsling commented on Source/JavaScriptCore/runtime/IntlLocale.cpp:
```diff
@@ -226,6 +226,28 @@ void IntlLocale::initializeLocale(JSGlobalObject* globalObject, JSValue tagValue
 }
 
+// https://tc39.es/proposal-intl-locale-info/#sec-weekday-to-string
+static StringView weekdayToString(StringView fw)
```
Oops, I didn't really mean to have you change the return type; I think this will work, since it's taking views of static string data, but ASCIILiteral would be most correct return type.

sosukesuzuki replied:
Why I changed the return type from `String` to `StringView` is as follows:
- The argument `fw` is of type `StringView`, so returning a `String` type requires an explicit conversion (the `weekdayToString` function returns `fw` as is when `fw` is neither `"0"` nor `"7"`).
- The return value of the `weekdayToString` function is passed as an argument to the functions `isUnicodeLocaleIdentifierType` and `localeID.setKeywordValue`. Both of these functions accept arguments of type `StringView`.

rkirsling replied:
Ah, I forgot about the `else` case.

sosukesuzuki committed:
[JSC] Implement `firstDayOfWeek` for `Intl.Locale` info API
https://bugs.webkit.org/show_bug.cgi?id=276943

Reviewed by Ross Kirsling.

JSC has implemented the Stage 3 Intl.Locale Info API[1][2]. The proposal includes support for the
"fw" (First day of the week) Unicode extension[3], but it is not currently implemented in JSC.

This patch implements support for the "fw" extension in the Intl.Locale Info API.

[1]: https://github.com/tc39/proposal-intl-locale-info
[2]: https://tc39.es/proposal-intl-locale-info
[3]: https://unicode.org/reports/tr35/#Key_Type_Definitions

* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/IntlLocale.cpp:
(JSC::weekDayToString):
(JSC::IntlLocale::initializeLocale):
(JSC::IntlLocale::firstDayOfWeek):
* Source/JavaScriptCore/runtime/IntlLocale.h:
* Source/JavaScriptCore/runtime/IntlLocalePrototype.cpp:
(JSC::JSC_DEFINE_CUSTOM_GETTER):

Canonical link: https://commits.webkit.org/281510@main
